---
title: "A tour of euPredictR"
author: "Tony Xie"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction_euPredictR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```



## Introduction

**euPredictR** is an R package for predicting protein-coding gene sequences in novel eukaryotic organisms based on sequence homology, using raw output generated by the Basic Local Alignment Search Tool (BLAST) algorithm.

To download `euPredictR`, use the following commands:

``` r
require("devtools")
devtools::install_github("tonyxieuoft/euPredictR", build_vignettes = TRUE)
library("euPredictR")
```

Before we describe the workflow of the package, let me first describe the biological context of the tool and specific terminology that I will use throughout this tutorial. 

The Basic Local Alignment Search Tool **(BLAST)** identifies high-similarity regions between a
_query_ and a _subject_ sequence. (Altschul et al. 1990) Each shared region of high similarity
between a query and subject is known as a “high-scoring pair” **(HSP)** and is associated with a
score known as an _E-value_. euPredictR analyzes raw BLAST output in specific cases where the
query sequences are protein-coding nucleotide sequences (aka. _coding sequences_ or _CDS_) for
genes from some species _A_, and the subject sequences are contiguous reads of a genome from
a different species _B_ whose coding sequence for the gene has not yet been determined. In this
context, the parts of the subject sequence within the regions of high similarity returned by
BLAST are considered by euPredictR as potential coding regions comprising the gene for
species _B_. This is also the approach that many use to predict gene sequences by sequence
homology for novel species (She et al. 2011; Keilwagen et al. 2018).

Note that eukaryotic coding sequences are composed of exon regions (which are separated by
introns in the direct genomic sequence). As a result, for a single coding sequence query against
a subject genomic sequence, BLAST may return many relevant high-scoring pairs that each
correspond to an exon (Wheeler and Bhagwat 2007). Additionally, spurious matches, gene
duplications, and other phenomena may result in many irrelevant high-scoring pairs (Wheeler
and Bhagwat 2007). The large number of high-scoring pairs make the process of separately
relevant from irrelevant ones complex. Scaling the analysis from the output of BLAST runs on
single queries (predicting the coding sequence for one gene) to many queries (predictions for
multiple genes, which is the anticipated use case) makes the process even more complicated.

Taking in raw output from BLAST runs on input coding sequence queries and subject genomic
sequences, `euPredictR` determines relevant high-scoring pairs, sort out overlaps, and constructs new, predicted coding sequence(s) from them.


## General Functionality Overview


`euPredictR` currently contains 5 functions available to the user.

1.   `parse_BLAST_json` parses a single raw BLAST output file provided 
in .json format. The function stores the parsed results in a nested list and data frame format for ease of access.

2.   `parse_multiple_BLAST_json` for parsing multiple raw BLAST output 
.json files in a given directory.

3.   `build_predictions` for creating protein-coding sequence (a.k.a coding sequence or CDS) prediction models for genes of particular species, based on parsed BLAST results. 

4.   `gene_coverage_heatmap` for visualizing gene coverage/prediction 
completeness of predicted coding sequences using a heatmap. 

5.   `output_predictions` for outputting the gene predictions in FASTA 
format.


## An Example

*Note - I wasn't able to get the knitting to work (the vignette can't seem to find my package), so in the meantime I have just included the skeleton for what the vignette should look like before I successfully troubleshoot it. 

Suppose that we have experimentally obtained protein-coding sequences from the bottlenose dolphin (T. truncatus) for 5 genes: IL10, IL6, TNF, RHO and SOX2. Let's suppose that we also _want_ to obtain the coding sequences of these 5 genes for four other species: the lamprey (P. marinus), mouse (M. musculus), orca (O. orca) and zebrafish (D. rerio). Unfortunately, in our fictitous example, no transcript sequences for them seem to be available anywhere. (Obviously, this isn't the case in reality, but it is useful for demonstrating the functionality of the package).

Luckily, let's suppose we have access to the genome assemblies of each species. Then, even though we don't actually have access to the coding sequences, we can predict them via sequence homology and extract them from the genome assemblies via BLAST. Since we only have coding sequences for the dolphin, we use them as queries to BLAST each subject genome assembly separately, and obtain raw BLAST output files in .json format, one per species. We also decide to BLAST the dolphin query coding sequences against the dolphin genome as a control (theoretically, the predicted coding sequences should be exactly the same as the query). Each raw BLAST output file contains a ton of data though, with HSPs that might not be relevant. Here's where *euPredictR* comes to the rescue.

Suppose all of our raw BLAST output files are in one directory. First, we want to name each file based on the species that the data represents. As a result, we get an identical directory to the "example_raw_output_directory" directory in the `./inst/extdata` folder. Then, we can run the `parse_multiple_blast_json()` via the following:

```r
dir_path <- "/path/to/blast/output/dir"
parsed_output <- euPredictR::parse_multiple_blast_json(dir_path = dir_path, 
                                                       raw_blast_list = NULL)
```

We used `parse_multiple_blast_json` because there were multiple BLAST output files to parse, and all were contained in one directory. The `raw_blast_list` parameter takes in a previously parsed output to concatenate with. Since this is the first BLAST .json file that we have parsed, we leave it empty. (see the help documentation for more).

Checking the type of our parsed output, we see:
```r
class(parsed_output)
```
that it is of type "RawBlastList". Each RawBlastList object is a nested list in key-value format, in which the outer list's keys are species and the values are 'gene lists' that are also lists in key-value format. The keys of 'gene lists' are gene names, while the values are lists of size 2 that contain the length of the query sequence used, as well as the HSP table produced for the query (and particular genomic assembly from one species). To find the HSP table for the RHO gene for the Orca reference genome, we see:
```r
parsed_output$O_orca$RHO$hsp_data
```
and to find the length of the query used to obtain the data, we see:
```r
parsed_output$O_orca$RHO$query_len
```
Now, with our parsed output, we can create gene predictions with the following:
```r
predictions <- euPredictR::build_predictions(raw_blast_list = parsed_output,
                                             max_overlap = 30,
                                             max_intron_length = 50000)
```
The `max_overlap` attribute is the maximum possible overlap between the query segments of two compatible HSPs, whereas the `max_intron_length` attribute is the maximum distance between the subject segments of two compatible HSPs.Read the help documentation for more. Checking the type of our predictions, we see:
```r
class(predictions)
```
that it is of type GenePredictions. To access the predicted sequence for Orca rhodopsin now (just like before), we see:







```r
#devtools::load_all(.)
```

```{r}
#sample_raw_blast_list

```


## References

* Altschul S.F., Gish, W., Miller, W., Myers, E.W. and Lipman, D.J. (1990). Basic local alignment
search tool. Journal of Molecular Biology 215(3): 403-410.
https://doi.org/10.1016/S0022-2836(05)80360-2

* Keilwagen, J., Hartung, F., Paulini, M., Twardziok, S.O., and Grau, J. (2018). Combining
RNA-seq data and homology-based gene prediction for plants, animals and fungi. BMC
Bioinformatics 19: e189. https://doi.org/10.1186/s12859-018-2203-5

* She, R., Chu, J.S., Uyar, B., Wang, J., Wang, K., Chen, N.S. (2011). genBlastG: using BLAST
searches to build homologous gene models. https://doi.org/10.1093/bioinformatics/btz714

* Wheeler, D. and Bhagwat, M. Blast QuickStart: Example-Driven Web-Based BLAST Tutorial. In:
Bergman NH, editor. Comparative Genomics: Volumes 1 and 2. Totowa (NJ): Humana Press; 2007. Chapter 9. https://www.ncbi.nlm.nih.gov/books/NBK1734/






