#' Display Phylogeny for a Single Gene
#'
#' Given an input nested list of predicted coding sequences in GenePredictions
#' format and an user-specified gene, construct a phylogeny from
#' predicted coding sequences for that single gene. After isolating the
#' relevant coding sequences for the gene from the
#' GenePredictions object, the sequences are then aligned with ClustalW,
#' a multiple-sequence alignment algorithm. Finally, a phylogeny is
#' inferred via the Unweighted Pair Group Method with Arithmetic Mean (UPGMA)
#' algorithm. UPGMA assumes that:
#' \itemize{
#' \item all taxa evolve at a constant rate, and that
#' \item all species also have the same mutation rate
#' }
#' Note that these assumptions are breached quite often in reality. Still,
#' although other methods like maximum-likelihood-based approaches have greater
#' accuracy, they take far longer to execute. The user should therefore treat
#' this function as a quick way to visually discern similarity between the
#' predicted coding sequences, rather than a real reflection of phylogenetic
#' relationships. If species appear closely related in the phylogenies produced,
#' the user should merely assume that their predicted sequences are closest to
#' each other in comparison to the sequences predicted for other species.
#'
#' @param predictions S3 object of class GenePredictions produced by an
#' iteration of the 'build_predictions' function in this package.
#' @param gene_name Name of gene to construct phylogeny from. As a requirement,
#' there must be at least coding sequence for the specified gene in
#' 'predictions'.
#'
#' @returns A plot of type "phylogram", generated by the base graphics package.
#' The user should use this plot as a quick way to discern similarity between
#' the predicted sequences of different species, rather than a real phylogenetic
#' tree.
#'
#' @examples
#'
#' # make gene predictions from sample RawBlastList data provided in package
#' predictions <- build_predictions(sample_raw_blast_list,
#'                                  max_overlap = 30,
#'                                  max_intron_length = 50000)
#'
#' # Example 1: Display phylogeny based on predicted sequences for RHO
#' \dontrun{
#'
#' display_phylogeny(predictions = predictions, gene_name = "RHO")
#'
#' }
#'
#' # Example 2: Display phylogeny based on predicted sequences for IL6
#' \dontrun{
#'
#' display_phylogeny(predictions = predictions, gene_name = "IL6")
#' # Note that the species that the algorithm failed to predict sequences for
#' # (P. marinus and D. rerio) are not in the phylogeny.
#'
#' }
#'
#' @import msa
#' @import dplyr
#' @import purrr
#' @import phangorn
#' @importFrom Biostrings DNAStringSet
#'
#' @export
#'
#' @references
#' Bodenhofer U, Bonatesta E, Horejs-Kainrath C, and Hochreiter S. (2015).
#' msa: an R package for multiple sequence alignment. Bioinformatics
#' 31(24):3997-9999. DOI: 10.1093/bioinformatics/btv176.
#'
#' Pagès H, Aboyoun P, Gentleman R, DebRoy S (2024). Biostrings: Efficient
#' manipulation of biological strings. doi:10.18129/B9.bioc.Biostrings.
#'
#' Schlie K, Potts AJ, Morrison DA, Grimm, GW. (2017), Intertwining
#' phylogenetic trees and networks. Methods in Ecology and Evolution, 8:
#' 1212--1220. doi:10.1111/2041-210X.12760
#'
#' Sokal R, Michener CD. (1958). A Statistical Method for Evaluating
#' Systematic Relationships. University of Kansas Scientific Bulletin
#' 28:1409-1438.
#'
#' Wickham H, François R, Henry L, Müller K, Vaughan D. (2023). dplyr: A
#' Grammar of Data Manipulation. R package version 1.1.4.
#' https://CRAN.R-project.org/package=dplyr.
#'
#' Wickham H, Henry L. (2023). purrr: Functional Programming Tools. R package
#' version 1.0.2. https://CRAN.R-project.org/package=purrr.
#'
display_phylogeny <- function(predictions, gene_name){

  if (!inherits(predictions, "GenePredictions")){
    stop("Incorrect input: 'predictions' must be in GenePredictions
         format and generated by the function build_predictions")
  }
  else{} # continue

  if (!is.character(gene_name) | length(gene_name) > 1){
    stop("Incorrect input: 'gene_name' must be a single string.")
  }
  else{} # continue

  # We get the total, unique gene names by the following:
  # 1) lapply applies the names function to each item in 'predictions'
  # to get the gene names present for each species
  # 2) reduce applies the union operator on the gene names until all
  # are concatenated into a single vector
  total_gene_names <- purrr::reduce(base::lapply(predictions, names),
                                    base::union)

  if (! gene_name %in% total_gene_names){
    stop("Incorrect input: The string specified to the parameter
         'gene_name' does not correspond to any gene predictions
         in the GenePredictions object passed to the 'predictions'
         parameter.")
  }
  else{} # continue

  # instantiates an empty vector to carry the sequences to be used in the
  # phylogeny (at most, there is one sequence for a gene per species)
  phylo_seqs <- rep(NA, times = length(predictions))

  species_names <- names(predictions)
  # names of each index correspond to each species name in the predictions
  names(phylo_seqs) = species_names

  # iterate through the gene predictions, extracting sequences that pertain
  # to the specific gene
  for (i in seq_along(predictions)){

    seq <- predictions[[i]][[gene_name]]

    if (! is.null(seq)){
      phylo_seqs[i] = seq
    }
    else{} # do nothing, keep the index for the species blank to remove later
  }

  # remove species that don't have a predicted coding sequence for the gene of
  # interest
  phylo_seqs <- phylo_seqs[!is.na(phylo_seqs)]

  # convert the vector of strings to DNAString format (necessary input format)
  dna_set <- Biostrings::DNAStringSet(phylo_seqs)

  # catch messages produced so they don't clutter the console for the user
  base::sink(file = base::nullfile())

  # get phylogeny by first aligning sequences with ClustalW
  phylo_plot <- suppressMessages(msa::msa(dna_set, "ClustalW")) %>%
    msaConvert("phangorn::phyDat") %>%   # convert to phangorn's phyDat format
    dist.ml() %>%                        # construct distance matrix
    phangorn::upgma() %>%                # conduct upgma calculations
    graphics::plot(main=paste("UPGMA phylogeny based on predicted",
                              gene_name, "sequences",
                              sep = " "))

  # reset the sink so print statements can be made again
  base::sink()

  # return the plot
  return(phylo_plot)
}

# [END]
